util/cmp/stream.go

```diff
@@ -0,0 +1,234 @@
+	}
+	metadata := header.GetMetadata()
+	workDir, err := ioutil.TempDir(os.TempDir(), metadata.GetAppName())
+	if err != nil {
+		return "", nil, fmt.Errorf("error creating workDir: %s", err)
+	}
+
+	tgzFile, err := receiveFile(ctx, receiver, metadata.GetChecksum(), workDir)
+	if err != nil {
+		return "", nil, fmt.Errorf("error receiving tgz file: %s", err)
+	}
+	err = files.Untgz(workDir, tgzFile)
+	if err != nil {
+		return "", nil, fmt.Errorf("error decompressing tgz file: %s", err)
+	}
+	err = os.Remove(tgzFile.Name())
+	if err != nil {
+		log.Warnf("error removing the tgz file %q: %s", tgzFile.Name(), err)
+	}
+	return workDir, metadata.GetEnv(), nil
+}
+
+// SendApplicationStream will compress the files under the given appPath and send
+// them using the plugin stream sender.
+func SendApplicationStream(ctx context.Context, appPath string, sender StreamSender, env []string) error {
+
+	// compress all files in appPath in tgz
+	tgz, checksum, err := compressFiles(appPath)
+	if err != nil {
+		return fmt.Errorf("error compressing app files: %s", err)
+	}
+	defer func() {
+		tgz.Close()
+		os.Remove(tgz.Name())
+	}()
+
+	// send metadata first
+	mr := appMetadataRequest(appPath, env, checksum)
+	err = sender.Send(mr)
+	if err != nil {
+		return fmt.Errorf("error sending generate manifest metadata to cmp-server: %s", err)
+	}
+
+	// send the compressed file
+	err = sendFile(ctx, sender, tgz)
+	if err != nil {
+		return fmt.Errorf("error sending app files to cmp-server: %s", err)
+	}
+	return nil
+}
+
+// sendFile will send the file over the gRPC stream using a
+// buffer.
+func sendFile(ctx context.Context, sender StreamSender, file *os.File) error {
+	reader := bufio.NewReader(file)
+	chunk := make([]byte, 1024)
+	for {
+		if ctx != nil {
+			if err := ctx.Err(); err != nil {
+				return fmt.Errorf("client stream context error: %s", err)
+			}
+		}
+		n, err := reader.Read(chunk)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("buffer reader error: %s", err)
+		}
+		fr := appFileRequest(chunk[:n])
+		err = sender.Send(fr)
+		if err != nil {
+			return fmt.Errorf("error sending stream: %s", err)
+		}
+	}
+	return nil
+}
+
+// compressFiles will create a tgz file with all contents of appPath
+// directory excluding the .git folder. Returns the file alongside
+// its sha256 hash to be used as checksum. It is the resposibility
+// of the caller to close the file.
+func compressFiles(appPath string) (*os.File, string, error) {
+	excluded := []string{".git"}
+	appName := filepath.Base(appPath)
+	tgzFile, err := ioutil.TempFile(os.TempDir(), appName)
+	if err != nil {
+		return nil, "", fmt.Errorf("error creating app temp tgz file: %s", err)
+	}
+	hasher := sha256.New()
+	err = files.Tgz(appPath, excluded, tgzFile, hasher)
+	if err != nil {
+		tgzFile.Close()
+		os.Remove(tgzFile.Name())
+		return nil, "", fmt.Errorf("error creating app tgz file: %s", err)
+	}
+	checksum := hex.EncodeToString(hasher.Sum(nil))
+
```

Comment:
I wonder if we should call `hasher.Reset()` here. If there happen to be any secrets in the hash state, resetting would just make sure those secrets are immediately removed from memory rather than waiting on GC.

Reply:
`hash.Hash` is an interface that just allows writes (no reads). The `hasher` variable is never returned by the function which makes it eligible for GC upon function conclusion.